File: SeoulWomen/app/src/main/java/kr/susemi99/seoulwomen/adapters/ClassListAdapter.java
package kr.susemi99.seoulwomen.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import kr.susemi99.seoulwomen.R;
import kr.susemi99.seoulwomen.models.RowItem;

public class ClassListAdapter extends RecyclerView.Adapter<ClassListAdapter.RowViewHolder> {
  private ArrayList<RowItem> items = new ArrayList<>();
  private View.OnClickListener clickListener;

  public ClassListAdapter(View.OnClickListener clickListener) {
    this.clickListener = clickListener;
  }

  public void add(RowItem item) {
    items.add(item);
  }

  public void clear() {
    items.clear();
    notifyDataSetChanged();
  }

  @Override
  public RowViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    int resId = R.layout.list_item_class;
    View view = LayoutInflater.from(parent.getContext()).inflate(resId, null);
    view
      .setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
    return new RowViewHolder(view);
  }

  @Override
  public void onBindViewHolder(RowViewHolder holder, int position) {
    RowItem item = items.get(position);
    holder.textDifficulty.setText("[" + item.difficultyName + "]");
    holder.textClassName.setText(item.className);
    holder.textReceiveAt.setText(item.displayReceiveFrom() + " " + item.receiveTimeFrom + " ~ " + item
      .displayReceiveTo() + " " + item.receiveTimeTo);
    holder.textEducateOn.setText(item.displayEducateFrom() + " ~ " + item.displayEducateTo());
    holder.textEducateAt.setText(item.displayDays() + " " + item.educateTimeFrom + " ~ " + item.educateTimeTo);
    holder.textSpare.setText(item.displaySpareNum() + "/" + item.displayCollectNum() + "명");
    holder.textFee.setText(item.displayEducateFee());
    holder.textRegist.setText(item.displayHowToRegist());

    holder.itemView.setTag(item.url);
    holder.itemView.setOnClickListener(clickListener);
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  public class RowViewHolder extends RecyclerView.ViewHolder {
    public TextView textDifficulty;
    public TextView textClassName;
    public TextView textReceiveAt;
    public TextView textEducateOn;
    public TextView textEducateAt;
    public TextView textSpare;
    public TextView textFee;
    public TextView textRegist;

    public RowViewHolder(View itemView) {
      super(itemView);
      textDifficulty = (TextView) itemView.findViewById(R.id.text_difficulty);
      textClassName = (TextView) itemView.findViewById(R.id.text_class_name);
      textReceiveAt = (TextView) itemView.findViewById(R.id.text_receive_at);
      textEducateOn = (TextView) itemView.findViewById(R.id.text_educate_on);
      textEducateAt = (TextView) itemView.findViewById(R.id.text_educate_at);
      textSpare = (TextView) itemView.findViewById(R.id.text_spare);
      textFee = (TextView) itemView.findViewById(R.id.text_fee);
      textRegist = (TextView) itemView.findViewById(R.id.text_how_to_regist);
    }
  }
}

File: SeoulWomen/app/src/main/java/kr/susemi99/seoulwomen/ui/ClassListAdapter.kt
package kr.susemi99.seoulwomen.ui

import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kr.susemi99.seoulwomen.application.App
import kr.susemi99.seoulwomen.databinding.ClassListItemBinding
import kr.susemi99.seoulwomen.model.RowItem

class ClassListAdapter : RecyclerView.Adapter<ClassListAdapter.AppViewHolder>() {
  private lateinit var binding: ClassListItemBinding

  private val items = arrayListOf<RowItem>()

  fun addAll(rowItems: List<RowItem>) {
    items.addAll(rowItems)
    notifyItemInserted(itemCount)
  }

  fun clear() {
    items.clear()
    notifyDataSetChanged()
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
    binding = ClassListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
    return AppViewHolder(binding.root)
  }

  override fun getItemCount(): Int {
    return items.size
  }

  override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
    val item = items[position]
    holder.bind(item)
    holder.itemView.setOnClickListener { goToHomePage(item) }
  }

  private fun goToHomePage(item: RowItem) {
    val uri = Uri.parse(item.url)
    val intent = Intent(Intent.ACTION_VIEW, uri).apply { addFlags(FLAG_ACTIVITY_NEW_TASK) }
    App.instance.startActivity(intent)
  }

  inner class AppViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    fun bind(item: RowItem) {
      itemView.apply {
        binding.difficultyLabel.text = "[${item.difficultyName}]"
        binding.classNameLabel.text = item.className
        binding.receiveAtLabel.text = "${item.displayReceiveFrom()} ${item.receiveTimeFrom} ~\n${item.displayReceiveTo()} ${item.receiveTimeTo}"
        binding.educateOnLabel.text = "${item.displayEducateFrom()} ~\n${item.displayEducateTo()}"
        binding.educateAtLabel.text = "${item.displayDays()}\n${item.educateTimeFrom} ~ ${item.educateTimeTo}"
        binding.spareLabel.text = "${item.displaySpareNum()}/${item.displayCollectNum()}명"
        binding.feeLabel.text = "${item.displayEducateFee()}"
        binding.howToRegistLabel.text = "${item.displayHowToRegist()}"
      }
    }
  }
}